### PR TITLE
feat(events): tracks events - base adapter version

### PR DESCRIPTION
**Pull request by dan-ziv**

### CODE REVIEW

File: src/engines/html5/media-source/adapters/native-adapter.js [hunk 1]
```diff
@@ -1,57 +1,50 @@
...
-   */
-  _config: Object;
-  /**
-   * The source object
-   * @member {Source} _sourceObj
-   * @private
-   */
-  _sourceObj: ?Source;
+  static set name(name: string): void {
+    // Do nothing. Just a workaround for flow issue with static getter in an inheritor. See: https://github.com/facebook/flow/issues/3008.
```
Comment (yairans) [inline src/engines/html5/media-source/adapters/native-adapter.js]: do we still need this workaround?

File: src/engines/html5/media-source/adapters/native-adapter.js [hunk 2]
```diff
@@ -75,46 +68,21 @@ export default class NativeAdapter implements IMediaSourceAdapter {
...
    * @param {Source} source - The source object
    * @param {Object} config - The media source adapter configuration
    */
   constructor(videoElement: HTMLVideoElement, source: Source, config: Object) {
-    this._config = config;
-    this._videoElement = videoElement;
-    this._sourceObj = source;
+    NativeAdapter._logger.debug('Creating adapter');
+    super(videoElement, source, config);
     this._eventManager = new EventManager();
```
Comment (yairans) [inline src/engines/html5/media-source/adapters/native-adapter.js]: should be in the base. no?

File: src/engines/html5/media-source/adapters/native-adapter.js [hunk 3]
```diff
@@ -153,6 +122,7 @@ export default class NativeAdapter implements IMediaSourceAdapter {
   destroy(): void {
     NativeAdapter._logger.debug('destroy');
     this._eventManager.destroy();
+    this._sourceObj = null;
```
Comment (yairans) [inline src/engines/html5/media-source/adapters/native-adapter.js]: should be in the base

File: src/engines/html5/media-source/adapters/native-adapter.js [hunk 4]
```diff
@@ -245,51 +215,48 @@ export default class NativeAdapter implements IMediaSourceAdapter {
...
    * @public
    */
-  selectVideoTrack(videoTrack: VideoTrack): boolean {
+  selectVideoTrack(videoTrack: VideoTrack): void {
     let videoTracks = this._videoElement.videoTracks;
     if ((videoTrack instanceof VideoTrack) && videoTracks && videoTracks[videoTrack.index]) {
       this._disableVideoTracks();
       videoTracks[videoTrack.index].selected = true;
-      return true;
+      this._trigger(BaseMediaSourceAdapter.CustomEvents.VIDEO_TRACK_CHANGED, {selectedVideoTrack: videoTrack});
```
Comment (yairans) [inline src/engines/html5/media-source/adapters/native-adapter.js]: should be in the base

File: src/engines/html5/media-source/adapters/native-adapter.js [hunk 5]
```diff
@@ -245,51 +215,48 @@ export default class NativeAdapter implements IMediaSourceAdapter {
...
    * @public
    */
-  selectAudioTrack(audioTrack: AudioTrack): boolean {
+  selectAudioTrack(audioTrack: AudioTrack): void {
     let audioTracks = this._videoElement.audioTracks;
     if ((audioTrack instanceof AudioTrack) && audioTracks && audioTracks[audioTrack.index]) {
       this._disableAudioTracks();
       audioTracks[audioTrack.index].enabled = true;
-      return true;
+      this._trigger(BaseMediaSourceAdapter.CustomEvents.AUDIO_TRACK_CHANGED, {selectedAudioTrack: audioTrack});
```
Comment (yairans) [inline src/engines/html5/media-source/adapters/native-adapter.js]: should be in the base

File: src/engines/html5/media-source/adapters/native-adapter.js [hunk 6]
```diff
@@ -245,51 +215,48 @@ export default class NativeAdapter implements IMediaSourceAdapter {
...
    * @public
    */
-  selectTextTrack(textTrack: TextTrack): boolean {
+  selectTextTrack(textTrack: TextTrack): void {
     let textTracks = this._videoElement.textTracks;
     if ((textTrack instanceof TextTrack) && (textTrack.kind === 'subtitles' || textTrack.kind === 'captions') && textTracks && textTracks[textTrack.index]) {
       this._disableTextTracks();
       textTracks[textTrack.index].mode = 'showing';
-      return true;
+      this._trigger(BaseMediaSourceAdapter.CustomEvents.TEXT_TRACK_CHANGED, {selectedTextTrack: textTrack});
```
Comment (yairans) [inline src/engines/html5/media-source/adapters/native-adapter.js]: should be in the base

File: src/engines/html5/media-source/base-media-source-adapter.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  static CustomEvents: { [event: string]: string } = CUSTOM_EVENTS;
+
+  static getLogger: Function = LoggerFactory.getLogger;
+
+  static get name(): string {
+    throw new PlayerError(PlayerError.TYPE.NOT_IMPLEMENTED_METHOD, 'get name').getError();
+  }
+
+  static set name(name: string): void {
+    // Do nothing. Just a workaround for flow issue with static getter in an inheritor. See: https://github.com/facebook/flow/issues/3008.
```
Comment (yairans) [inline src/engines/html5/media-source/base-media-source-adapter.js]: still necessary?

File: src/engines/html5/media-source/adapters/native-adapter.js (hunk 2 above)
Comment (dan-ziv) [inline src/engines/html5/media-source/adapters/native-adapter.js]: What should be in the base? please explain.
if you are talking about the event manager, so no, hls and dash wont need him.

File: src/engines/html5/media-source/adapters/native-adapter.js (hunk 4 above)
Comment (dan-ziv) [inline src/engines/html5/media-source/adapters/native-adapter.js]: what should be in the base?
the selection itself is an adapter unique logic.
the event trigger is in the base.
not sure I understand you here.

File: src/engines/html5/media-source/adapters/native-adapter.js (hunk 5 above)
Comment (dan-ziv) [inline src/engines/html5/media-source/adapters/native-adapter.js]: what should be in the base?
the selection itself is an adapter unique logic.
the event trigger is in the base.
not sure I understand you here.

File: src/engines/html5/media-source/adapters/native-adapter.js (hunk 6 above)
Comment (dan-ziv) [inline src/engines/html5/media-source/adapters/native-adapter.js]: what should be in the base?
the selection itself is an adapter unique logic.
the event trigger is in the base.
not sure I understand you here.

File: src/engines/html5/media-source/adapters/native-adapter.js (hunk 1 above)
Comment (dan-ziv) [inline src/engines/html5/media-source/adapters/native-adapter.js]: done.
(no need getter and setter at all actually, just to define _name in the derived adapter)

**Pull request merged**